Add OpenTelemetry Operator Helm chart to the existing Helm repo (#52)
open-telemetry/opentelemetry-helm-charts: charts/opentelemetry-operator/README.md

metadata:
  name: my-collector
spec:
  mode: statefulset
  volumeMounts:
    - mountPath: "/usr/share/test-volume"
      name: test-volume
  volumeClaimTemplates:
    - metadata:
        name: "test-volume"
      spec:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 1Gi
  replicas: 3
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
    processors:

    exporters:
      logging:

    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: []
          exporters: [logging]
EOF
```

### Sidecar Mode
If all you want to monitor is a single application, then collector as the sidecar would be the best fit.
This collector instance will work on the container level and no new pods or other workload resources are needed,
which is perfect to keep your Kubernetes cluster clean. Moreover, you can also use the sidecar mode when you want
to use a different collect/export strategy, which just suits this application.

You can deploy the sidecar mode by setting the pod annotation `sidecar.opentelemetry.io/inject` to either `true`,

> or to the name of a concrete `OpenTelemetryCollector` from the same namespace.

Author: If there is only one collector with the mode `sidercar` in a specific namespace, then users could just set `sidecar.opentelemetry.io/inject` to true. But if there are multiple collectors with the mode `sidecar` in the same namespace, then users much provide the collector pod name. It should be more clear here. I'll change the expression.

$ helm repo update
```

_See [helm repo](https://helm.sh/docs/helm/helm_repo/) for command documentation._

## Install Chart

The following command installs the chart with the release name `my-opentelemetry-operator` in the namespace `opentelemetry-operator-system`.
You don't need to provide `--namespace` option here.
```console
$ helm install \
  my-opentelemetry-operator open-telemetry/opentelemetry-operator
```

_See [helm install](https://helm.sh/docs/helm/helm_install/) for command documentation._

## Uninstall Chart

The following command uninstalls the chart whose release name is my-opentelemetry-operator.

```console
$ helm uninstall my-opentelemetry-operator
```

_See [helm uninstall](https://helm.sh/docs/helm/helm_uninstall/) for command documentation._

This will remove all the Kubernetes components associated with the chart and deletes the release.

The OpenTelemetry Collector CRD created by this chart won't be removed by default and should be manually deleted:

```console
$ kubectl delete crd opentelemetrycollectors.opentelemetry.io
```

The namespace created should also be removed manually:

```console
$ kubectl delete ns opentelemetry-operator-system

> ```

Reviewer: Since you defined namespace in the yaml instead of `install ... --create-namespace`, this step is not needed anymore, namespace will be removed with `helm uninstall`

Author: Yes, I overlooked that. I will remove this step. Thank you!

# OpenTelemetry Operator Helm Chart

The Helm chart installs [OpenTelemetry Operator](https://github.com/open-telemetry/opentelemetry-operator) in Kubernetes cluster.
The OpenTelemetry Operator is an implementation of a [Kubernetes Operator](https://www.openshift.com/learn/topics/operators).
At this point, it has [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector) as the only managed component.

## Prerequisites

> - Kubernetes 1.19+

Author: It's the [operator requirement](https://github.com/open-telemetry/opentelemetry-operator/blob/main/README.md#opentelemetry-operator-vs-kubernetes).

> - Kubernetes 1.19+

Reviewer 2: Sorry, I should have made it clear :-) The section should tell users that this is an operator requirement.

> - Kubernetes 1.19+

Author: Got it! I will make the change.

- Helm 3.0+

### TLS Certificate Requirement

In Kubernetes, in order for the API server to communicate with the webhook component, the webhook requires a TLS
certificate that the API server is configured to trust. There are three ways for you to generate the required TLS certificate.

>   - The easiest and default method is to install the cert-manager and keeps `admissionWebhooks.certManager.enabled` to `true`.

Author: Yeah for sure!

    In this way, cert-manager will generate a self-signed certificate. _See [cert-manager installation](https://cert-manager.io/docs/installation/kubernetes/) for the instruction._
  - You can also provide your own Issuer by configuring the `admissionWebhooks.certManager.issuerRef` value. You will need
    to specify the `kind` (Issuer or ClusterIssuer) and the `name`. Note that this method also requires the installation of cert-manager.
  - The last way is to manually modify the secret where the TLS certificate is stored. You can either do this before installation
    or after.
    - To do this before installation, you don't have to install cert-manager. Please set `admissionWebhooks.certManager.enabled` to `false`.

>       - Create the namespace for the OTEL Operator and the secret

Reviewer 2: ```suggestion
      - Create the namespace for the OpenTelemetry Operator and the secret
```

        ```console
        $ kubectl create namespace opentelemetry-operator-system
        ```
      - Config the TLS certificate
        ```console
        $ kubectl apply -f - <<EOF
          apiVersion: v1
          kind: Secret
          metadata:
            name: opentelemetry-operator-controller-manager-service-cert
            namespace: opentelemetry-operator-system
          type: kubernetes.io/tls
          data:
            tls.crt: |
                # your signed cert
            tls.key: |
                # your private key
          EOF
        ```

>         You can also do this using `kubectl create` command.

Reviewer 2: This is the preferred option. I'd recommend listing this first. Or even removing the one above.

Author: Yeah, sure. I will make the changes.

        ```console
        $ kubectl create secret tls opentelemetry-operator-controller-manager-service-cert \
            --cert=path/to/cert/file \
            --key=path/to/key/file \
            -n opentelemetry-operator-system
        ```

>     - To do this after installation, you will have to install cert-manager first. Once the Operator is ready, you can override

Reviewer 2: Is this supported by cert-manager? Won't cert-manager override this when the cert is up for renewal?

Author: I tried to set a very short `duration` in `Certificate` resource and change the secret after the operator installation. It turns out cert-manager will override this when the cert it up for renewal. So, the current way doesn't work.

I think if the users want to use their own cert after the operator installation, they could use `kubectl edit` command to edit the `Certificate` resource (e.g. `kubectl edit cert opentelemetry-operator-serving-cert -n opentelemetry-operator-system`) and replace the `issuerRef` with their own CA. Or, we can just remove this from the README. Please let me know what you think. Thank you!

Reviewer 2: Just remove this from the readme. We should give them one way of doing things. If they need to do it differently, they will find a way. Using `kubectl create secret tls` should be the one, IMO.

Author: This makes sense.

      the secret by putting your certificate there. What you need to do is the same as the second step above.

## Add Repository

```console
$ helm repo add open-telemetry https://open-telemetry.github.io/opentelemetry-helm-charts
$ helm repo update
```

_See [helm repo](https://helm.sh/docs/helm/helm_repo/) for command documentation._

## Install Chart

If you didn't create the namespace `opentelemetry-operator-system` before (steps in the third method to generate the TLS cert),

> the OTEL Operator chart will be installed in the namespace automatically. Installation command example is as below.

Reviewer 2: ```suggestion
the OpenTelemetry Operator chart will be installed in the namespace automatically. The installation command example is as below.
```

```console
$ helm install \
  my-opentelemetry-operator open-telemetry/opentelemetry-operator
```

However, if you create the namespace and place the TLS cert in the desired secret, you will need to set `createNamespace`
to `false` to make sure Helm won't try to create an existing namespace, which would cause an error. Installation command example is as below.

```console
$ helm install \
  my-opentelemetry-operator open-telemetry/opentelemetry-operator \
  --set createNamespace=false
```

> Note that `--namespace` option here won't affect on where the OTEL Operator and other resources this chart contains are installed.

Reviewer 2: ```suggestion
Note that `--namespace` option here won't affect where the OpenTelemetry Operator and other resources this chart contains are installed.
```

It will only affect on where the Helm chart release info is stored, which is `default` namespace by default.

_See [helm install](https://helm.sh/docs/helm/helm_install/) for command documentation._

## Uninstall Chart

The following command uninstalls the chart whose release name is my-opentelemetry-operator.

```console
$ helm uninstall my-opentelemetry-operator
```

_See [helm uninstall](https://helm.sh/docs/helm/helm_uninstall/) for command documentation._

This will remove all the Kubernetes components associated with the chart and deletes the release.

The OpenTelemetry Collector CRD created by this chart won't be removed by default and should be manually deleted:

```console
$ kubectl delete crd opentelemetrycollectors.opentelemetry.io
```

> If the namespace is created by Helm instead of yourself, please skip this step. If you created the namespace and set

Reviewer 2: This paragraph is a bit confusing. Perhaps simplify it? 

> If the namespace wasn't created by the Helm chart, you'll need to manually remove it as well:

Author: Thanks!

```

Please note that by default, the chart will be upgraded to the latest version. If you want to upgrade to a specific version,
use `--version` flag.

With Helm v3.0, CRDs created by this chart are not updated by default and should be manually updated.
Consult also the [Helm Documentation on CRDs](https://helm.sh/docs/chart_best_practices/custom_resource_definitions).

_See [helm upgrade](https://helm.sh/docs/helm/helm_upgrade/) for command documentation._

## Configuration

The following command will show all the configurable options with detailed comments.

```console
$ helm show values open-telemetry/opentelemetry-operator
```

## Install OpenTelemetry Collector

_See [OpenTelemetry website](https://opentelemetry.io/docs/collector/) for more details about the Collector_

Once the opentelemetry-operator deployment is ready, you can deploy OpenTelemetry Collector in our Kubernetes
cluster.

The Collector can be deployed as one of four modes: Deployment, DaemonSet, StatefulSet and Sidecar. The default
mode is Deployment. We will introduce the benefits and use cases of each mode as well as giving an example for each.

### Deployment Mode

> If you want to get more control of the OpenTelemetry collector and create a standalone application, Deployment would

Reviewer 2: ```suggestion
If you want to get more control of the OpenTelemetry Collector and create a standalone application, Deployment would
```

> be your choice. With Deployment, you can relatively easily scale up the collector to monitor more targets, roll back

Reviewer 2: ```suggestion
be your choice. With Deployment, you can relatively easily scale up the Collector to monitor more targets, roll back
```

to an early version if anything unexpected happens, pause the collector, etc. In general, you can manage your collector
instance just as an application.

> The following example configuration deploys the Collector as Deployment resource. The receiver is jaeger-receiver and

Reviewer 2: You have two styles of naming components: `jaeger-receiver` and `logging exporter`. Either use `Jaeger receiver` and `logging exporter`, or `jaeger-receiver` and `logging-exporter`. This last option seems strange to me though.

Author: I will unify them using the former one.

the exporter is logging exporter.

```console
$ kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: my-collector
spec:
  mode: deployment # This configuration is omittable.
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
    processors:

    exporters:
      logging:

    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: []
          exporters: [logging]
EOF
```

### DaemonSet Mode

> DaemonSet should satisfy your most basic needs and is the most common method to deploy OpenTelemetry collector.

Reviewer 2: ```suggestion
DaemonSet should satisfy your most basic needs and is the most common method to deploy OpenTelemetry Collector.
```

Is it, though? I would love to hear more about it :-)

Author: Here is my thought - please correct me if I'm wrong. Installing Collector as DaemonSet to the k8s cluster basically means you are installing an agent and every k8s node has its own Collector copy. And I guess the most common needs would be using the Collector as an agent to scrape the telemetry data. Just my 2 cents.

Reviewer 2: My contention is that you are claiming that daemonsets are the most common method of deploying the collector. I don't think I've seen data about it, but if you can back it, I'm all for keeping this. Note that the OpenTelemetry isn't only about scraping data, but also about receiving data, especially when we talk about traces. In that respect, sidecars are common and popular, as they are a buffer between the workload and the remote collector, and an outage in the collector at the node level won't affect all pods on that node.

Author: "the most common method" is absolute. I will rephrase this statement to clarify that in what situation, DaemonSet would be a good choice.

Yes, as you said, Collector has multiple receivers to ingest data. We should be clear to users what mode is best for what.

> In this case, every Kubernetes node will have its own collector copy which would monitor the pods in it.

Reviewer 2: ```suggestion
In this case, every Kubernetes node will have its own Collector copy which would monitor the pods in it.
```

> The following example configuration deploys the Collector as DaemonSet resource. The receiver is jaeger-receiver and

Reviewer 2: Same as an earlier comment.

the exporter is logging exporter.

```console
$ kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: my-collector
spec:
  mode: daemonset
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
    processors:

    exporters:
      logging:

    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: []
          exporters: [logging]
EOF
```

### StatefulSet Mode

> If you want your collector to have stable persistent identities or storage, you should choose StatefulSet.

Reviewer 2: ```suggestion
If you want your Collector to have stable persistent identities or storage, you should choose StatefulSet.
```

Persistent storage isn't a promise of StatefulSets. When a pod fails, Kubernetes will _attempt_ to schedule the new pod with the same name to the node that held the pod until now. Kubernetes will also _attempt_ to mount the same volume it was used before.

This means that if the node is down, if the pod with the same name will be created in a different node. When using local storage, this means that data will be lost.

Author: > This means that if the node is down, if the pod with the same name will be created in a different node. When using local storage, this means that data will be lost.

I'm sorry, but can you elaborate on this? I'm not sure if I get your points. Thanks.

Reviewer 2: Sure, let me try, hoping that I have had enough coffee this time :-) When having a stateful set with one replica, you'll have a pod named "otelcol-0". This will be scheduled to run in the node, say, "node-0".

When using a local storage like hostPath, when "otelcol-0" crashes, Kubernetes will attempt to schedule the new "otelcol-0" to the same node (node-0). If Kubernetes can't schedule the pod to this node (perhaps the node itself crashed), then a new node will be assigned to the pod. Now, "otelcol-0" is running on "node-1". Your hostPath now points to a location that has no data at all, so, your state has not been persisted.

Hence my comment that persistent storage isn't a promise of StatefulSets. They will attempt to run on the same node and attach the same volumes, but using a statefulset doesn't guarantee that you'll see the same data as before when a pod is restarted.

Author: Thanks for this explanation. The scenario that the node itself may crash should be considered.

Since in the Collector case, persistent storage isn't what's the StatefulSet focuses, I removed the PV related expression and underscored three advantages instead. Please LMK what's your thoughts on that.

  If a Collector pod fails in the StatefulSet, Kubernetes will attempt to reschedule a new pod with the same name to the same node. Kubernetes will also attempt
  to attach the same sticky identity (e.g., volumes) to the new pod.

> - The load balancer could be configured (under construction) \

Reviewer 2: I would leave this point off until the load balancer is available. Is it going to be part of the operator logic, or will it be part of a collector component?

Author: Yeah sure, I will remove this for now. It's going to be part of the operator logic. There are two PRs and we have already filed the first one, as you may see.

> interface called PersistentVolume. PersistentVolume will maintain the historical metrics data and even survive pods

Reviewer 2: Persistent volumes can be used with deployments as well. The difference is that volumes attached to pods that are part of a stateful set will always be the same (unless they are local volumes). With deployments, a new claim will be made for each new pod.

If the collector is not writing anything to the local storage (as is usually the case, as the collector is typically stateless), the only advantage of stateful sets is the predictable naming.

Going back to your example: are you saying that the Prometheus receiver is locally storing data on disk? I thought it would scrape the endpoints and dispatch the data points through the pipeline.

Author: I agree in our case, the StatefulSet Collector doesn't have the advantage in terms of the volumes because it doesn't write data to the local storage. But I think other than predictable names, loadbalancing (design doc [here](https://docs.google.com/document/d/1GdpCqdkz72291l_8S-jio7EhVltminobvqshsFih6vA/edit?usp=sharing), implementation is ongoing) is another advantage which the StatefulSet Collector owns. Maybe I could remove the persistent storage part and give a few descriptions about the loadbalancing advantages here. When this feature is merged upstream, I will add more details.

> are you saying that the Prometheus receiver is locally storing data on disk? I thought it would scrape the endpoints and dispatch the data points through the pipeline.

No, sorry for the confusion. I was trying to give the [Prometheus metrics local storage examples](https://prometheus.io/docs/prometheus/latest/storage/) to better illustrate the idea of the PV. I will rephrase this paragraph to be more clear.

restart. This feature gives you the possibility that you can reconstruct the time-series logs.

> The following example configuration deploys the Collector as StatefulSet resource with three replicas. The receiver

Reviewer 2: Same as an earlier comment.

The following example configuration deploys the Collector as StatefulSet resource with three replicas. The receiver
is Jaeger receiver and the exporter is logging exporter.

```console
$ kubectl apply -f - <<EOF
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: my-collector
spec:
  mode: statefulset

>   volumeMounts:

Reviewer 2: The collector isn't using this volume in the configuration. This isn't needed, is it?

Author: Yes, this configuration is neglectable since the Collector doesn't use it at this time.

  volumeClaimTemplates:
    - metadata:
        name: "test-volume"
      spec:
        accessModes: [ "ReadWriteOnce" ]
        storageClassName: "standard"
        resources:
          requests:
            storage: 1Gi
  replicas: 3
  config: |
    receivers:
      jaeger:
        protocols:
          grpc:
    processors:

    exporters:
      logging:

    service:
      pipelines:
        traces:
          receivers: [jaeger]
          processors: []
          exporters: [logging]
EOF
```

### Sidecar Mode
If all you want to monitor is a single application, then collector as the sidecar would be the best fit.
This collector instance will work on the container level and no new pods or other workload resources are needed,
which is perfect to keep your Kubernetes cluster clean. Moreover, you can also use the sidecar mode when you want

> to use a different collect/export strategy, which just suits this application.

Reviewer 2: While this is certainly a valid case, I wouldn't say this is the biggest use case. This mode exists for people to offload their telemetry data as fast and reliable as possible from their applications. A sidecar (ie, container in the same pod) is good for that, as it's on the same network as the application's process. 

This sidecar would typically send the telemetry data to a remote OpenTelemetry Collector, either as a stateful set or deployment, which is then configured to send data to a final destination. If you decide to switch your Jaegers or to send data to a new vendor, you can just change the collector configuration in the deployment. Changing the sidecars would require all the pods to be recreated.

Author: > This mode exists for people to offload their telemetry data as fast and reliable as possible from their applications. A sidecar (ie, container in the same pod) is good for that, as it's on the same network as the application's process.

Will add this to the README!

You can deploy the sidecar mode by setting the pod annotation `sidecar.opentelemetry.io/inject` to `true` if there is only one
collector with mode `sidecar` in the namespace. Otherwise, you will need to set that value to the name of a concrete

> `OpenTelemetryCollector` pod from the namespace.

Reviewer 2: > Once a Sidecar instance exists in a given namespace, you can have your deployments from that namespace to get a sidecar by either adding the annotation `sidecar.opentelemetry.io/inject=true` to the pod spec of your application, or to the namespace.

## Install Chart

The following command installs the chart with the release name `my-opentelemetry-operator` in the namespace `opentelemetry-operator-system`.

> You don't need to provide `--namespace` option here.

Reviewer: I'd move [the note](https://github.com/open-telemetry/opentelemetry-helm-charts/pull/52/files/7f82e64f0ae10f0e78baebfbe0eaef4d2dc5f000..b4eae4235d6f8ce988476039f95fd9fb9c9787b6#diff-c9849889db5d4f8fc6aa865982eb9bece6cd4bb31fa6286972d5327413a00955R7-R9) here and say that `--namespace` helm argument doesn't have any effect on where the kubernetes objects will be installed. User still can set `--namespace` to put helm release info in any other namespace, if not specified it'll be `default` which is fine I guess.

Author: I will rephrase the README.
